### ISSUE TITLE
Error with Normalize and TabularPandas: causes 09_tabular.ipynb to fail

### ISSUE THREAD

**Issue by juliangilbey** [fastai/fastai#3231]
The details are described in https://github.com/fastai/fastbook/issues/411

A bit more investigating has shown this strange behaviour: at the point where `encodes` is called, `self` is pretty empty; this is the result of printing `self` in `fastai/tabular/core.py` in the `@Normalize encodes` function (lines 276 onwards):
```python
self = Normalize -- {'mean': None, 'std': None, 'axes': (0, 2, 3), 'means': {}, 'stds': {}}:
encodes: (TensorImage,object) -> encodes
(Tabular,object) -> encodes
decodes: (TensorImage,object) -> decodes
(Tabular,object) -> decodes
```
So it looks as though the `store_attr` call in `setups` was not effective.  Printing out `self` in `setups` after the `store_attr` call suggests this as well.

I have to admit that the decorators are also completely flooring me from debugging much further: I don't understand how `@Normalize` in `fastai/tabular/core.py` actually works or why the code path even ends up here, especially as `setups` is redefined several times in the file, and `Normalize` does not take a function as argument in its `__init__`.  But anyway, there's something clearly not working correctly here. :(

**Comment by juliangilbey** [comment]
OK, I've located the problem - it's in the `09_tabular.ipynb` notebook itself.  So closing this report.

**Comment by juliangilbey** [comment]
Ah, no, it turns out there are errors in both `09_tabular.ipynb` and in `fastai`.  So reopening this with the diagnosis and a patch for the `fastai` library part.

In https://github.com/fastai/fastai/blob/master/nbs/40_tabular.core.ipynb, we have the definition of `add_datepart`:
```python
#export
def add_datepart(df, field_name, prefix=None, drop=True, time=False):
    "Helper function that adds columns relevant to a date in the column `field_name` of `df`."
    make_date(df, field_name)
    field = df[field_name]
    prefix = ifnone(prefix, re.sub('[Dd]ate$', '', field_name))
    attr = ['Year', 'Month', 'Week', 'Day', 'Dayofweek', 'Dayofyear', 'Is_month_end', 'Is_month_start',
            'Is_quarter_end', 'Is_quarter_start', 'Is_year_end', 'Is_year_start']
    if time: attr = attr + ['Hour', 'Minute', 'Second']
    # Pandas removed `dt.week` in v1.1.10
    week = field.dt.isocalendar().week.astype(field.dt.day.dtype) if hasattr(field.dt, 'isocalendar') else field.dt.week
    for n in attr: df[prefix + n] = getattr(field.dt, n.lower()) if n != 'Week' else week
    mask = ~field.isna()
    df[prefix + 'Elapsed'] = np.where(mask,field.values.astype(np.int64) // 10 ** 9,None)
    if drop: df.drop(field_name, axis=1, inplace=True)
    return df
```

Now, the intention of the `df[prefix + 'Elapsed']` is to add an `int64` column with the elapsed time (time since long ago or whatever).  Unfortunately, though, the `np.where` function returns an `ndarray` whose `dtype` is the smallest that can accommodate both arguments, in this case an array of type `int64` and `None`, and that is `O` ("object").  So the resulting Elapsed column has type "object", which is not particularly useful.

Easy fix: add an extra line after this assignment:
```python
    df[prefix + 'Elapsed'] = df[prefix + 'Elapsed'].astype('Int64')
```

(This cannot be combined into a single line, as `np.where` returns a NumPy `ndarray`, but `Int64` is a Pandas type, so can only be applied to a Pandas Series or DataFrame.  Therefore the assignment into `df` needs to happen before the change of dtype.)

**Comment by aberres** [comment]
I stumbled upon the same problem.

The other solution might be to switch to a float dtype - as this is a continuous variable this is the correct thing to do IMO. I proposed this MR: https://github.com/fastai/fastai/pull/3230

@jph00 needs to decide, I guess.

**Comment by juliangilbey** [comment]
Ah, if we're changing to a float dtype, then the solution is even simpler: just change the 'Elapsed' definition line to read:
```python
    df[prefix + 'Elapsed'] = np.where(mask,field.values.astype(np.int64) / 1e9,np.nan)
```
The `np.nan` is a `float` type, so this ends up with the whole column having dtype `float64`.  (And we can replace `// 10 ** 9` with `/ 1e9` because it's going to be converted into a float anyway.)

**Comment by jph00** [comment]
Is this resolved now?

**Comment by juliangilbey** [comment]
We have two possible resolutions, but it's your choice which to apply to the code base: either the suggestion to add the extra line
```python
    df[prefix + 'Elapsed'] = df[prefix + 'Elapsed'].astype('Int64')
```
after the assignment (which requires a relatively recent Pandas, I think >= 0.25), which keeps it an integer type, or replace the assignment with
```python
    df[prefix + 'Elapsed'] = np.where(mask,field.values.astype(np.int64) / 1e9,np.nan)
```
which makes the Elapsed column a float64 instead.

**Comment by aberres** [comment]
@juliangilbey Have a look at https://github.com/fastai/fastai/pull/3230 (released in v2.2.7) and https://github.com/fastai/fastbook/pull/413

I think this resolves the problem.

**Comment by juliangilbey** [comment]
@jph00 @aberres Ah, I hadn't seen that - thanks for the fix.  Yes, this resolves the problem.  (The only small thing, probably of negligible significance, is that if times are given with millisecond accuracy, this will be lost by the integer division.  Changing `// 10 ** 9` into `/ 1e9` would address this, too.)
